### DISCUSSION TITLE
Datagrid + Bulk Actions

### DISCUSSION THREAD

**Discussion by ceyerkes** [twilio-labs/paste#3942] [General] [answered]
Hello! I'm working on a pretty complex data grid for the new alerting platform, and i'm wondering if we have any docs/established patterns around bulk actions (delete, edit) within a data grid. happy to bring to office hours, just let me know, thank you!

![Screenshot 2024-06-17 at 9 21 41 AM](https://github.com/twilio-labs/paste/assets/87141551/f1f517d6-b2d6-4f2a-a691-f4490bd201cc)

**Comment by roxanagomez** [comment]
Hi @ceyerkes thank you for sharing, we don't have a specific pattern for bulk actions in our roadmap, it would be great to [get started on a contribution](https://paste.twilio.design/patterns/contribute/) if you'd be interested in that.  Also, two questions:

- Do you know if there are other teams that might be working on this pattern that you could collaborate with?
- Do you have a Figma file exploration or some guidelines for the pattern?

It be good if we have some specific points to discuss during office hours, but feel free to sign up if you want to just bounce ideas of.

**Reply by ceyerkes** [reply]
Yes, I'd be happy to contribute at some point.

I don't know if anyone else is working on a similar pattern, but I could ask in Slack.
I don't have guidelines for this pattern, just the [figma file](https://www.figma.com/design/bHTlaFk9EARwTjoPAcj9WD/Alerting-%5BDesign%5D?node-id=3244-169411&t=BwG5tHjXYnVg7TI9-4) that i'm working in where I'm working through options that fit our specific use case.

**Reply by jess-watson** [reply]
I'm exploring a similar bulk actions pattern in the journeys workflow builder – here's the [figma file](https://www.figma.com/design/JXc2nHCFyg8V4VZFE1RQJF/Journeys-Canvas-UX?node-id=137-169236&t=rblkN3R8cms4guHs-1).

**Reply by kkenna3** [reply]
Voice has use cases built or being built, and there are other upcoming use cases for Console. Documenting the existing UI & use cases here [Figma](https://www.figma.com/design/qKGSuEr7euXXvxcKy8MGW6/Data-Grid-Bulk-Actions?node-id=0-1&t=K37PwKarzoHLJLYH-1 )

**Reply by roxanagomez** [reply]
Great, I think the next steps for this would be to:

1. Start with a rough draft of guidelines, even bullet points will do. It can be figma or in docs. 
2. Cut scope aggressively. The pattern doesn't need to cover every use case to start.
3. Give us a Figma page or component set of the ideal flow/interaction, even if you're not 100% confident about it—as long as you're more confident about it than the alternatives. 

We're happy to provide feedback at any point during office hours if needed.

**Reply by ceyerkes** [reply]
Awesome, sounds good! I'm out for a couple weeks after the 4th, but will touch base with @jess-watson and @kkenna3 when I'm back.

**Reply by elangager3** [reply]
This is also something that I have been looking into, please let me know how I can help contribute in any way!

**Reply by ceyerkes** [reply]
@kkenna3 @jess-watson and @elangager3 - i'll set up a meeting for us to go over @roxanagomez's next steps outlined above, and we can figure out how we want to tackle this contribution, if that sounds good.

**Comment by nkrantz** [comment, marked as answer]
Hey @ceyerkes - thanks for kicking off this contribution in office hours today! You can find the recording in the #help-design-system channel.

Did you think more about how it might work to conditionally display the available actions in the Data Grid itself? Happy to collaborate on options here if it's something you want to move forward with.

**Reply by ceyerkes** [reply]
So sorry for the late reply, @nkrantz!! After chatting with @serifluous, I'm thinking for this iteration, we might consider keeping it super simple, and not changing anything within the UI of the data grid, but adjust the options in the actions menu to reflect what's available for single vs bulk actions (disabling options only available for single rows). 

I also would like to explore the option of adding "undo" to a toast after a bulk action is performed:

![Screenshot 2024-10-22 at 9 17 55 AM](https://github.com/user-attachments/assets/1605023d-bc79-4ece-850e-c8b68d5429a4)

Happy to come to office hours again and chat more - this solves my use case for now, but I'm open to continuing the convo to explore a more complicated bulk action pattern.

**Reply by nkrantz** [reply]
Great! I'll mark this discussion as answered for now but feel free to post updates here in the future. [Here's the section of Toast docs with an "undo" example](https://paste.twilio.design/components/toast#success-toast) - just a reminder that Toasts with interactive elements as children shouldn't time out.

**Reply by ceyerkes** [reply]
Ahhhh i didn't know that existed for Toasts! Awesome, thank you Nora!
